adds a preprocessor rule to add IDs to block data

jroebu14 opened the pull request:
Resolves (partially) https://github.com/bbc/simorgh/issues/1489

**Overall change:** Adds unique IDs to block data so that React renders blocks more performantly.

**Code changes:**

- Adds a new preprocessor rule `addIdsToBlocks` to recursively add IDs to article block data.
- Replaces nanoid usage in Blocks container with IDs generated in `addIdsToBlocks` as block keys.
- Replaces nanoid usage in ErrorPage using sanitised error strings as list item keys.
- Updates `article.test.js` to ensure `fetchData` is called with `addIdsToBlocks` preprocessor rule.

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Tests added for new features
- [x] Test engineer approval
- [x] I have followed [the merging checklist](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/MERGE_PROCESS.md) and this is ready to merge.

Commit:
adds a preprocessor rule to add IDs to block data

ryanmccombe committed:
Replace spread operators with deep cloning

ryanmccombe committed:
Merge branch 'latest' of github.com:bbc/simorgh into remove-nanoid-keys

ryanmccombe committed:
Fix unit test

ryanmccombe commented:
@jroebu14 spread operator was actually handling nulls fine - the reason the tests were failing on the CI server was that that the article data had changed on the latest branch, and that property was removed

Possibly a good reason to not have this unit test rely on the test data so exactly - any change to that test data will break the unit test

I only realised the pre-processor was handling nulls fine after I refactored it :(  It should be good to go back to the original implementation if preferred

Commit:
separate functions and make declarative

Commit:
spread object properties outside of block content in case they change and break test

Commit:
Merge branch 'remove-nanoid-keys' of github.com:bbc/simorgh into remove-nanoid-keys

jroebu14 commented:
@ryanmccombe ah ok well spotted. I changed the test to spread parent properties of `blocks` so that if the article data does change then we don't need to change it in the test. It will still break if `blocks` data within the article changes but that's a good thing to catch in case we need to change the `addIdsToBlocks` preprocessor rule to cater for the change in the data structure.

I do prefer the first implementation of `addIdsToBlocks` because the style was more functional. I reverted it back and also changed it to be slightly more declarative. Let me know what you think.

Commit:
Merge branch 'latest' into remove-nanoid-keys

ryanmccombe committed:
Merge branch 'latest' into remove-nanoid-keys

Commit:
merges rules and default rules array in preprocessor

Commit:
ensures preprocessor rules are called only once

Commit:
Merge branch 'remove-nanoid-keys' of github.com:bbc/simorgh into remove-nanoid-keys

jroebu14 commented on src/app/lib/utilities/preprocessor/index.js:
```diff
@@ -1,11 +1,12 @@
 import nodeLogger from '../../../helpers/logger.node';
 import applyTimestampRules from './rules/timestamp';
 
-const defaultRules = [applyTimestampRules];
```
Hi @ChrisBAshton I had to change how default rules is used here. Do you think default rules and new passed in rules should work like the changes I made? Or was it your intention for new rules to completely override the default rules? If the latter then it would mean I would have to pass in the timestamp rule into https://github.com/bbc/simorgh/pull/1879/files#diff-9927c9cefc48863f54ee5f0b62a72ee4L12

jamesbhobbs replied:
@jroebu14 @ChrisBAshton no longer works for the beeb, but he's a really nice guy and might answer anyway - just don't count on it/wait for it :)

dr3 replied:
You could just get rid of default rules, they dont really make sense anyway, as any other page type that doesnt use article data will need to override them anyway.

Line 7 wont work for WS frontpages which cant use the `applyTimestampRules` rule as its a different data format

jroebu14 replied:
@dr3 OK this makes sense to us. I think we should remove `applyTimestampRules` and use it in `getArticleInitialData`. This makes `preprocessor` more generic too.

ryanmccombe replied:
Sounds good - one side effect is it does change the behaviour for any existing consumers - anyone who called `fetchData` without any preprocessor rules would have previously had the `applyTimestampRules` formatter applied.

If we remove `defaultRules`, they no longer will, but this is likely fine.  `applyTimestampRules` doesn't seem to make sense as a default in the long term

Commit:
Removes default rules. Uses timestamp rules in article fetch

Commit:
Fixes article test ensure timestamp rule is used

jamesbhobbs reviewed (changes requested):
Looks good just needs nanoid removed from package.json and packagelock

ryanmccombe committed:
Merge branch 'latest' into remove-nanoid-keys

jamesbhobbs reviewed (changes requested):
I was wrong that you're not using nanoid...
Ok so, if we're still using nanoid, I have to ask why. We are using uuid via deps, so we don't save on amounts of code by using it vs uuid. Also I am not sure this PR fixes the issue as the misuse of nanoid is also drawn through from here https://github.com/bbc/psammead/blob/fa3f52ec9a8aade35cecd0813822bad9572994d2/packages/components/psammead-sitewide-links/src/List/index.jsx isn't it? Suggest this latter part is fixed in another PR but it means this PR doesn't fix the issue.

Commit:
replaces nanoid package with uuid

Commit:
adds IDs to footer links

Commit:
Merge branch 'remove-nanoid-keys' of github.com:bbc/simorgh into remove-nanoid-keys

jroebu14 commented:
@benjaminhobbs nanoid package has been removed from dependencies and replaced with uuid. I've also added custom IDs to the links array elements which is passed into the SitewideLinks component. There will soon be a Psammead pull request to make use of the IDs and remove the nanoid usage there.

dr3 commented on src/app/containers/Footer/__snapshots__/index.test.jsx.snap:
```diff
@@ -17,30 +17,37 @@ exports[`FooterContainer should render correctly 1`] = `
       Array [
         Object {
           "href": "https://www.bbc.com/news/help-41670342",
+          "id": "trust",
```
Curious why you're using an id here, rather than the text, it seems a tad superfluous. And could also cause issues in future.

If you navigate from 1 service to another, when we localise this data, the text could change, but the key wont, confusing react.

jroebu14 replied:
@dr3 I suppose it is a tad superfluous. I'm sure we can be confident that we will never have 2 footer links with identical text. I'll remove the IDs here and use just the text as the key in the Psammead component.

Commit:
Removes IDs in footer links

Commit:
Updates footer snapshot

jamesbhobbs committed:
Merge branch 'latest' into remove-nanoid-keys

dr3 reviewed (dismissed):
LGTM,
Checked on real article and output is whats expected, nice work
![image](https://user-images.githubusercontent.com/11341355/59350968-3f756a80-8d15-11e9-9cb7-ff81388ff2f8.png)

Commit:
Merge branch 'latest' into remove-nanoid-keys

benjaminhobbs commented:
@jroebu14 needs a package lock file update

pjlee11 commented on src/app/lib/utilities/preprocessor/rules/addIdsToBlocks.js:
```diff
@@ -0,0 +1,48 @@
+
+const getJsonContent = jsonRaw => deepGet(['content'], jsonRaw);
+
+const getBlocks = content => deepGet(['model', 'blocks'], content);
```
```suggestion
const getBlocks = block => deepGet(['model', 'blocks'], block);
```

`content` is the name of the first block that we encounter in simorgh, however it is semantically just another block. Also because this function is recursively called `content` as a variable name has a different meaning when this function is called by a nested block.

pjlee11 commented on src/app/lib/utilities/preprocessor/rules/addIdsToBlocks.js:
```diff
@@ -0,0 +1,48 @@
+  const blockWithId = { ...block, id: uuid() };
+  const nestedBlocks = getBlocks(blockWithId);
+
+  return nestedBlocks
```
Personally I think ternary is more readable as a guard clause. 

```
  if (!nestedBlocks) {
    return blockWithId;
  }

  return {
    ...blockWithId,
    model: {
      ...blockWithId.model,
      blocks: mapIdsToBlocks(nestedBlocks),
    },
  };


```

pjlee11 reviewed (dismissed):
Couple of minor points, great work so far! 😎

Commit:
semantic and readabillity refactoring

Commit:
package-lock updates

Commit:
Merge branch 'latest' into remove-nanoid-keys

Commit:
Merge branch 'latest' into remove-nanoid-keys

Commit:
Merge branch 'remove-nanoid-keys' of github.com:bbc/simorgh into remove-nanoid-keys

Commit:
Merge branch 'latest' into remove-nanoid-keys

Commit:
Merge branch 'latest' into remove-nanoid-keys

Commit:
update preprocessor rules in article test

Commit:
Merge branch 'latest' into remove-nanoid-keys

Commit:
fix paths

Commit:
update package lock

amywalkerdev committed:
Merge branch 'latest' into remove-nanoid-keys

jamesbhobbs committed:
Merge branch 'latest' into remove-nanoid-keys

PriyaKR commented:
Regression looks good to me.

jamesbhobbs committed:
Merge branch 'latest' into remove-nanoid-keys
